### Question 1

In [464]:
import pandas as pd
import numpy as np
from sklearn import linear_model
from sklearn.model_selection import train_test_split

### Read dictionary.csv file

In [467]:
dict = pd.read_csv('dictionary.csv')

### Column names

In [470]:
print(dict.columns)

Index(['Country', 'Code', 'Population', 'GDP per Capita'], dtype='object')


### Shape

In [473]:
print(dict.shape)

(201, 4)


### Check for missing values

In [476]:
dict.isna().sum()

Country            0
Code               0
Population         5
GDP per Capita    25
dtype: int64

In [478]:
dict.head()

,Country,Code,Population,GDP per Capita
0,Afghanistan,AFG,32526562.0,594.323081
1,Albania,ALB,2889167.0,3945.217582
2,Algeria,ALG,39666519.0,4206.031232
3,American Samoa*,ASA,55538.0,NaN
4,Andorra,AND,70473.0,NaN


### Mean for Population and GDP per Capita

In [481]:
dict[['Population', 'GDP per Capita']].describe()

,Population,GDP per Capita
count,1.960000e+02,176.000000
mean,3.722825e+07,12882.556131
std,1.399655e+08,17747.141203
min,1.022200e+04,277.068309
25%,1.638278e+06,1781.096847
50%,7.450124e+06,5233.583395
75%,2.557454e+07,15494.683646
max,1.371220e+09,101449.968168


In [483]:
print(dict[['Population', 'GDP per Capita']].mean())

Population        3.722825e+07
GDP per Capita    1.288256e+04
dtype: float64


### Proposed method to fill the missing values
#### Since there are only 30 missing values, composing only 3.7% in this dataset of 801 values I would use dropna().
#### I do not believe the missing values would contribute significatly.

## Question 2: 
#### Use Car_Purchasing_Data.csv file, this file is in the Linear Regression folder. In the auto example, find the multi-linear relationship between 'Age','Annual Salary', 'Net Worth' with 'Car Purchase Amount'. 

In [487]:
auto = pd.read_csv('LinearReg/Car_Purchasing_Data.csv')

In [489]:
auto.head()

,Customer Name,Customer e-mail,Country,Gender,Age,Annual Salary,Credit Card Debt,Net Worth,Car Purchase Amount
0,Martina Avila,cubilia.Curae.Phasellus@quisaccumsanconvallis.edu,USA,0,42,62812.09301,11609.380910,238961.2505,35321.45877
1,Harlan Barnes,eu.dolor@diam.co.uk,USA,0,41,66646.89292,9572.957136,530973.9078,45115.52566
2,Naomi Rodriquez,vulputate.mauris.sagittis@ametconsectetueradip...,USA,1,43,53798.55112,11160.355060,638467.1773,42925.70921
3,Jade Cunningham,malesuada@dignissim.com,USA,1,58,79370.03798,14426.164850,548599.0524,67422.36313
4,Cedric Leach,felis.ullamcorper.viverra@egetmollislectus.net,USA,1,57,59729.15130,5358.712177,560304.0671,55915.46248


In [491]:
print(auto.columns)
print(auto.dtypes)

Index(['Customer Name', 'Customer e-mail', 'Country', 'Gender', 'Age',
       'Annual Salary', 'Credit Card Debt', 'Net Worth',
       'Car Purchase Amount'],
      dtype='object')
Customer Name           object
Customer e-mail         object
Country                 object
Gender                   int64
Age                      int64
Annual Salary          float64
Credit Card Debt       float64
Net Worth              float64
Car Purchase Amount    float64
dtype: object


In [493]:
# Dropping object columns

auto.drop(["Customer Name", "Customer e-mail", "Country"], \
          axis=1, inplace=True)
print(auto.shape)

print(auto.columns)

(500, 6)
Index(['Gender', 'Age', 'Annual Salary', 'Credit Card Debt', 'Net Worth',
       'Car Purchase Amount'],
      dtype='object')


In [495]:
# creating an instance of LinearRegression class
reg = linear_model.LinearRegression()
print(reg)

LinearRegression()


In [497]:
auto.describe()

,Gender,Age,Annual Salary,Credit Card Debt,Net Worth,Car Purchase Amount
count,500.000000,500.000000,500.000000,500.000000,500.000000,500.000000
mean,0.506000,46.224000,62127.239608,9607.645049,431475.713625,44209.799218
std,0.500465,7.990339,11703.378228,3489.187973,173536.756340,10773.178744
min,0.000000,20.000000,20000.000000,100.000000,20000.000000,9000.000000
25%,0.000000,41.000000,54391.977195,7397.515792,299824.195900,37629.896040
50%,1.000000,46.000000,62915.497035,9655.035568,426750.120650,43997.783390
75%,1.000000,52.000000,70117.862005,11798.867487,557324.478725,51254.709517
max,1.000000,70.000000,100000.000000,20000.000000,1000000.000000,80000.000000


In [499]:
auto.drop(['Gender', 'Credit Card Debt'], axis=1, inplace=True)

In [501]:
auto.describe()

,Age,Annual Salary,Net Worth,Car Purchase Amount
count,500.000000,500.000000,500.000000,500.000000
mean,46.224000,62127.239608,431475.713625,44209.799218
std,7.990339,11703.378228,173536.756340,10773.178744
min,20.000000,20000.000000,20000.000000,9000.000000
25%,41.000000,54391.977195,299824.195900,37629.896040
50%,46.000000,62915.497035,426750.120650,43997.783390
75%,52.000000,70117.862005,557324.478725,51254.709517
max,70.000000,100000.000000,1000000.000000,80000.000000


In [503]:
from sklearn.preprocessing import StandardScaler

In [505]:
a_scaler = StandardScaler() # creating an instance of Standard Scaler

In [507]:
auto_independent = a_scaler.fit_transform(auto[['Age', 'Annual Salary', 'Net Worth']])

In [509]:
print(type(auto_independent))

<class 'numpy.ndarray'>


In [511]:
auto[['Age', 'Annual Salary', 'Net Worth']].head()

,Age,Annual Salary,Net Worth
0,42,62812.09301,238961.2505
1,41,66646.89292,530973.9078
2,43,53798.55112,638467.1773
3,58,79370.03798,548599.0524
4,57,59729.15130,560304.0671


In [513]:
# after scaling
auto_independent[0:5]

array([[-0.52916784,  0.05857619, -1.11046945],
       [-0.65444432,  0.38657041,  0.57392937],
       [-0.40389136, -0.71236095,  1.19397625],
       [ 1.47525579,  1.47479358,  0.67559542],
       [ 1.34997932, -0.20511086,  0.74311275]])

In [515]:
# using train_test_split(), we are splitting the data into training and test 
# x_train, x_test, y_train, y_test

x_train, x_test, y_train, y_test = train_test_split(auto_independent, auto["Car Purchase Amount"], test_size=0.2, random_state=4)

In [517]:
# we have to fit the data
reg.fit(x_train, y_train)

LinearRegression()

In [519]:
print(reg.coef_) # 
print(reg.intercept_)

[6698.99994256 6569.701515   5019.75983599]
44203.833191718


In [521]:
yhat = reg.predict(x_test) 

### Mean Squared Error for test set

In [524]:
from sklearn.metrics import mean_squared_error

mse_test = mean_squared_error(y_test, yhat)
print(mse_test)

60987.694591014406


### Mean Squared Error for the train set

In [527]:
# y_t_predict is the predicted y values for the x_train data
y_t_predict = reg.predict(x_train)

# note that y_train is the true y value
mse_train = mean_squared_error(y_train, y_t_predict)
print(mse_train)

57696.853858354705


In [529]:
r1 = mse_test/mse_train

diff1 = np.abs((mse_train - mse_test)/mse_train)

print(r1)

print(diff1)

1.0570367448585445
0.05703674485854441


### r-squared for test and train set

In [532]:
from sklearn.metrics import r2_score

print("r-squared for the test data: ", r2_score(y_test, yhat))
    
print("r-squared for the train data: ", r2_score(y_train, y_t_predict))

r-squared for the test data:  0.9994323113012907
r-squared for the train data:  0.9995037057564383


#### Conclusion:

r-squared for test is 0.9994  and for train is 0.9995

The multi-linear model seems like a good fit.